File: scripts/csv2json.rb
require "csv"
require "json"
require "fileutils"

def csv2json(path)
  csv = CSV.read(path, headers: true, converters: :all)

  headers = csv.headers
  rows = csv
    .to_a
    .drop(1)
    .map{ |row| Hash[headers.zip(row)] }
    .reject{ |h| h.values.all? { |v| v.nil? } } # remove possible empty rows

  destination_path = [
    __dir__,
    "../app/assets/data",
    File.dirname(path).sub(/^.*data\//, ""),
    "#{File.basename(path, ".csv")}.json",
  ]
  json_path = File.expand_path File.join(*destination_path)
  json_dir = File.expand_path File.join(*destination_path[0..-2])

  # create directory if needed
  FileUtils.mkdir_p(json_dir) unless File.directory?(json_dir)

  File.open(json_path, "w") do |file|
    file.write(rows.to_json)
  end
end

path = ARGV[0]

if !File.directory?(path) and ((path || "").empty? or File.extname(path.strip) != ".csv")
  puts <<~HELP
    Invalid file or wrong extension

    Usage: ruby scripts/csv2json.rb path/to/folder
    or
    Usage: ruby scripts/csv2json.rb path/to/file.csv
  HELP

  return 1
end

if File.directory?(path)
  puts "Processing directory..."
  Dir.glob("#{path.sub(/\/$/, "")}/**/*.csv").each do |file|
    print "Processing #{file}... "
    # puts "#{file}: wrote #{csv2json(file)} bytes"
    csv2json(file)
    puts "Done!"
  end
else
  puts "Is single file"
  csv2json(path)
end

File: scripts/csv2json.js
/* eslint-disable no-console */

import path from "path";
import fs from "fs";
import glob from "glob";
import mkdirp from "mkdirp";
import parse from "csv-parse/lib/sync";

const geographiesKeys = [
  "geo2010_nf",
  "zip_nf",
  "anc2012_nf",
  "psa2012_nf",
  "ward2012_nf",
  "county_nf",
  "cluster_tr2000_nf",
];

// downcase all geography column names to ensure consistency
function transformColumns(columns) {
  return columns.map(column => (
    geographiesKeys.includes(column.toLowerCase())
      ? column.toLowerCase()
      : column
  ));
}

function csv2json(file, from, to) {
  const destination_path = path.join(...[
    process.cwd(),
    to,
    path.dirname(file).replace(from, ""),
  ]);

  const json_path = path.join(destination_path, `${path.basename(file, ".csv")}.json`).toLowerCase();
  const csv_path = path.join(destination_path, path.basename(file)).toLowerCase();

  mkdirp.sync(destination_path);

  console.log(`Processing ${file}:`);

  fs.createReadStream(file).pipe(fs.createWriteStream(csv_path));
  console.log(`Copied ${file} to ${csv_path}`);

  const json =
    parse(fs.readFileSync(file), { columns: transformColumns, auto_parse: true, skip_empty_lines: true })
    .map(row => {
      const overrides = {};

      if (row.timeframe) overrides.timeframe = row.timeframe.toString();

      return { ...row, ...overrides };
    });

  fs.writeFileSync(json_path, JSON.stringify(json));
  console.log(`Converted ${file} to ${json_path}\n`);
}

//-----------------------------------------------------------------------------
// Main script

const argv = require("minimist")(process.argv.slice(2));
const { from, to } = argv;

if (!from || !to) {
  console.log("\nMissing arguments");
  console.log("Usage: node scripts/csv2json.js --from=path/to/sources --to=path/to/destination\n");

  process.exit(1);
}

try {
  if (!fs.lstatSync(from).isDirectory()) throw new Error();
}
catch(e) {
  console.log(`Invalid path '${from}'`)

  process.exit(1);
}

glob(`${from.replace(/\/$/, "")}/**/*.csv`, (err, files) => {
  files.forEach(file => csv2json(file, from, to));
});
